spec/requests/api/v2/sessions/create_by_magic_spec.rb

```diff
@@ -0,0 +1,113 @@
+require "rails_helper"
+
+describe "POST /sessions/create_by_magic" do
+  let(:user)    { create(:user, :with_magic_link_token) }
+  let(:payload) { { email: email } }
+  let(:email)   { user.email }
+  let(:magic_link_token) { user.magic_link_token }
+  let(:params) { { magic_link_token: magic_link_token } }
+
+  let(:token) do
+    JWT.encode payload,
+               Settings.authentication.secret,
+               Settings.authentication.algorithm
+  end
+
+  let(:credentials) do
+    ActionController::HttpAuthentication::Token.encode_credentials(token)
+  end
+
+  def perform_request
+    patch(
+      create_by_magic_api_v2_sessions_path,
+      headers: { "HTTP_AUTHORIZATION" => credentials },
+      params: params,
+    )
+  end
+
+  context "token and email are correct" do
+    it "sets the user's token and sent_at to nil" do
+      perform_request
+
+      user.reload
+      expect(user.magic_link_token).to be_nil
+      expect(user.magic_link_token_sent_at).to be_nil
+    end
+
+    it "sets the user's last login date" do
+      perform_request
+
+      user.reload
+      expect(user.last_login_date_utc).to be_within(4.seconds).of Time.now.utc
+    end
+
+    it "records the first login" do
+      user.update(first_login_date_utc: nil)
+
+      perform_request
+
+      user.reload
+      expect(user.first_login_date_utc).to be_within(4.seconds).of Time.now.utc
+    end
+
+    it "sends a welcome email" do
+      allow(SendWelcomeEmailService).to receive(:call)
```

Comment:
We should check `ActionMailer::Base.deliveries` here so as we test our full stack rather than stubbing our service.

Reply:
Ya, I figured out how to do this for another test, forgot to back-port. Done.